packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp

```diff
@@ -3208,31 +3207,29 @@ namespace Ifpack2 {
       
     public:      
       void run(const impl_scalar_type_2d_view &zz,
-	       const local_ordinal_type blocksize,
 	       magnitude_type *vals) {
-	IFPACK2_BLOCKTRIDICONTAINER_PROFILER_REGION_BEGIN("BlockTriDi::ReduceMultiVector");
-
+	IFPACK2_BLOCKTRIDICONTAINER_PROFILER_REGION_BEGIN("BlockTriDi::ReduceMultiVector");	
 	const local_ordinal_type nrows = zz.extent(0);
 	const local_ordinal_type ncols = zz.extent(1);
 
-	/// currently the norm manager interface asks elementwise norm in a block multi vector
-	/// this is not necessary and we will address this later
-	/// right now evenly distribute the multivector nomr over block entries
-	if (ncols == 1) {
-	  const auto norm = KokkosBlas::nrm1(Kokkos::subview(zz, Kokkos::ALL(), 0))/magnitude_type(blocksize);
-	  for (local_ordinal_type i=0;i<blocksize;++i)
-	    vals[i] = norm;
-	} else {	  
-	  Kokkos::View<magnitude_type*,device_type> rv("rv", ncols);
-	  KokkosBlas::nrm1(rv, zz);
-	  const auto rv_host = Kokkos::create_mirror_view_and_copy
-	    (typename host_execution_space::memory_space(), rv);
-	  for (local_ordinal_type j=0;j<ncols;++j) {
-	    const auto norm = rv_host(j)/magnitude_type(blocksize);
-	    for (local_ordinal_type i=0;i<blocksize;++i)
-	      vals[j*blocksize+i] = norm;
-	  }
-	}
+	impl_scalar_type_1d_view rz(zz.data(), nrows*ncols);
+#if 0
+	// this is expensive when I use cublas with uvm memory...
+	// no clue yet.
+	const auto norm2 = KokkosBlas::nrm2(rz);
```

Comment:
Given that your hand-rolled kernel is faster and that `rz` has rank 1, two possibilities come to mind:

  1. `KokkosBlas::nrm2` is calling cuBLAS, and cuBLAS is slow.
  2. `KokkosBlas::nrm2` is not calling cuBLAS, and the `final` kernel that takes the square root (on device!) may be slow.  See `V_Nrm2_Functor`, in `kokkos-kernels/src/blas/impl/KokkosBlas1_nrm2_impl.hpp`.  Just passing `take_sqrt=false` into `KokkosBlas::nrm2` won't help, since the `final` kernel will always run (`take_sqrt` is a run-time `bool` in the kernel).

Reply:
Are you sure nrm2 is actually calling cuBLAS?

I've been looking at the CuBLAS/GEMM thing - (trying to figure out exactly what is happening on Cuda...)

GEMM (with y'all patches!) looks great: 4-5x improvement in total time!
```

        |   |       |   |   |   |   |-> 6.77e+01 sec 5.5% 5.2% 0.3% 28.5% 47990 Belos::MVT::MvTransMv [region]
        |   |       |   |   |   |   |   |-> 3.65e+01 sec 3.0% 0.0% 2.9% 5.2% 47990 Tpetra::MV::multiply [region]
        |   |       |   |   |   |   |   |   |-> 2.42e+01 sec 2.0% 0.0% 1.7% 4.5% 47990 Tpetra::MV::multiply-call-gemm [region]
        |   |       |   |   |   |   |   |   |   |-> 2.31e+01 sec 1.9% 0.0% 1.7% 100.0% 47990 KokkosBlas::gemm[TPL_BLAS,double] [region]
        |   |       |   |   |   |   |   |   |-> 1.04e+01 sec 0.8% 0.0% 13.0% 100.0% 47990 Tpetra::MV::reduce [region]
```

Notice GEMM is marked TPL_BLAS... norm2 is not.
```
 |   |   |   |   |   |-> 1.18e+01 sec 1.0% 32.5% 11.2% 58.1% 47990 Tpetra::MV::norm2 (Teuchos::ArrayView) [region]
        |   |       |   |   |   |   |       |-> 4.44e+00 sec 0.4% 75.0% 5.4% 25.0% 47990 KokkosBlas::nrm2[ETI] [region]
        |   |       |   |   |   |   |       |   |-> 3.33e+00 sec 0.3% 100.0% 6.5% ------ 47990 KokkosBlas::Nrm2::S0 [reduce]
```

Reply:
For my own testing, I changed kokkoskernels to eti uvm space. Without cublas, it shows identical performance and with cublas, it is slower. If it is faster, I would include the modification on the kokkoskernels in this PR. As it is slower in my test, I did not include it to prevent any side effect of other slowness it might cause.